espnet/nets/beam_search.py

```diff
@@ -278,32 +307,7 @@ def forward(self, x: torch.Tensor, maxlenratio: float = 0.0, minlenratio: float
         ended_hyps = []
         for i in range(maxlen):
             logging.debug('position ' + str(i))
-            best = []
-            for hyp in running_hyps:
-                scores, states = self.score(hyp, x)
-                part_ids = self.pre_beam(scores, device=x.device)
-                part_scores, part_states = self.score_partial(hyp, part_ids, x)
-
-                # weighted sum scores
-                weighted_scores = torch.zeros(self.n_vocab, dtype=x.dtype, device=x.device)
-                for k in self.full_scorers:
-                    weighted_scores += self.weights[k] * scores[k]
-                for k in self.part_scorers:
-                    weighted_scores[part_ids] += self.weights[k] * part_scores[k]
-                weighted_scores += hyp.score
-
-                # update hyps
-                for j, part_j in zip(*self.main_beam(weighted_scores, part_ids)):
-                    # will be (2 x beam at most)
-                    best.append(Hypothesis(
-                        score=(weighted_scores[j]),
-                        yseq=self.append_token(hyp.yseq, j),
-                        scores=self.merge_scores(hyp, scores, j, part_scores, part_j),
-                        states=self.merge_states(states, part_states, part_j)))
-
-                # sort and prune 2 x beam -> beam
-                best = self.top_beam_hyps(best)
-
+            best = self.search(running_hyps, x)
```

Comment:
I agree. I think that the best way for performance and backward compatibility is returning two values:
- `batch_running_hyps: BatchHypothesis`: a constant topk pool of running hypotheses to avoid memory reallocation overhead
- `ended_hyps: List[Hypothesis]` a new list of ended hypotheses to reuse existing post-processing functions